src/lib.rs

```diff
@@ -0,0 +1,353 @@
+/// giving us plenty of room for future PQC schemes:
+/// - `0x40` = ML-DSA 65
+/// - `0x41` = SLH-DSA SHA2 256 S
+#[derive(Debug, PartialEq, Clone, Copy)]
+pub enum PubKeyType {
+    /// ML-DSA 65 public key.
+    MLDSA65,
+    /// SLH-DSA SHA2 256 S public key.
+    SLHDSASHA2_256S,
+}
+
+impl PubKeyType {
+    /// Byte code to embed in the payload.
+    ///
+    /// Reserved range: 0x40..=0xFF (up to 192 public key types).
+    #[must_use]
+    pub fn code(self) -> u8 {
+        match self {
+            PubKeyType::MLDSA65 => 0x40,
+            PubKeyType::SLHDSASHA2_256S => 0x41,
+        }
+    }
+
+    /// Reverse lookup from byte code to enum.
+    #[must_use]
+    pub fn from_code(code: u8) -> Option<PubKeyType> {
+        match code {
+            0x40 => Some(PubKeyType::MLDSA65),
+            0x41 => Some(PubKeyType::SLHDSASHA2_256S),
+            _ => None,
+        }
+    }
+}
+
+/// The default hash function for pq-address: SHA-256.
+///
+/// 256 bit hash function is currently considered secure against Grover's attack.
+/// Even if the preimage is recovered, it only reveals a PQ secure public key and thus Shor's is not applicable.
+pub struct Hasher;
+
+impl Hasher {
+    /// SHA-256 digest length (bytes).
+    pub const DIGEST_LENGTH: usize = 32;
+
+    /// Compute SHA-256(data).
+    #[must_use]
+    pub fn digest(data: &[u8]) -> Vec<u8> {
+        let mut h = Sha256::new();
+        h.update(data);
+        h.finalize().to_vec()
+    }
+}
+
+/// Parameters needed to encode an address.
+pub struct AddressParams<'a> {
+    /// Mainnet vs Testnet
+    pub network: Network,
+    /// Which address version
+    pub version: Version,
+    /// Which public‑key type
+    pub pubkey_type: PubKeyType,
+    /// Raw public‑key bytes
+    pub pubkey_bytes: &'a [u8],
+}
+
+/// Errors that can occur during address encoding.
+#[derive(Error, Debug)]
+pub enum AddressEncodeError {
+    /// Invalid Bech32 structure or checksum
+    #[error("Bech32 error: {0}")]
+    Bech32(#[from] bech32::EncodeError),
+
+    /// A Bech32 string is at most 90 characters long [BIP-173]
+    #[error("A Bech32 string is at most 90 characters long: got {0}")]
+    InvalidEncodingLength(usize),
+}
+
+/// Encodes a Bech32m address from given params.
+///
+/// # Errors
+/// Returns a `bech32::EncodeError` if the encode step fails.
+pub fn encode_address(params: &AddressParams) -> Result<String, AddressEncodeError> {
+    // Hash the public key
+    let digest = Hasher::digest(params.pubkey_bytes);
+
+    // Build the raw payload:
+    //  [ version.code(), pubkey_type.code(), digest bytes… ]
+    let mut payload = Vec::with_capacity(2 + digest.len());
+    payload.push(params.version.code());
+    payload.push(params.pubkey_type.code());
+    payload.extend(&digest);
+
+    // Bech32m‑encode (adds the 6‑word checksum)
+    let hrp = params.network.hrp();
+    let encoded = encode::<Bech32m>(hrp, &payload)?;
+
+    if encoded.len() > MAX_ADDRESS_LENGTH {
```

Comment:
We could be more specific here and check the exact length of the encoding if we wanted, considering it should always be the same